packages/flutter_tools/lib/src/build_system/build_system.dart

```diff
@@ -0,0 +1,581 @@
+    @required this.copyDir,
+    @required this.targetPlatform,
+    @required this.buildMode,
+  });
+
+  /// The `PROJECT_DIR` magic environment varaible.
+  final Directory projectDir;
+
+  /// The `STAMP_DIR` magic environment variable.
+  ///
+  /// Defaults to `{PROJECT_ROOT}/build`.
+  final Directory stampDir;
+
+  /// The `BUILD_DIR` magic environment variable.
+  ///
+  /// Defaults to `{PROJECT_ROOT}/build`.
+  final Directory buildDir;
+
+  /// The `CACHE_DIR` magic environment variable.
+  ///
+  /// Defaults to `{FLUTTER_ROOT}/bin/cache`.
+  final Directory cacheDir;
+
+  /// The `COPY_DIR` magic environment variable.
+  ///
+  /// Defaults to `{PROJECT_ROOT}/{host_folder}/flutter`
+  final Directory copyDir;
+
+  /// The currently selected build mode.
+  final BuildMode buildMode;
+
+  /// The current target platform, or `null` if none.
+  final TargetPlatform targetPlatform;
+}
+
+class BuildSystem {
+  const BuildSystem([this.targets = allTargets]);
+
+  final List<Target> targets;
+
+  /// Build the target `name` and all of its dependencies.
+  Future<void> build(
+    String name,
+    Environment environment,
+  ) async {
+    // Initialize any destination directories that don't currently exist.
+    if (!environment.cacheDir.existsSync()) {
+      environment.cacheDir.createSync(recursive: true);
+    }
+    if (!environment.copyDir.existsSync()) {
+      environment.copyDir.createSync(recursive: true);
+    }
+
+    // Compute the required order of targets.
+    final List<Target> ordered = _computeTargetOrder(targets, name, environment);
+
+    // Visit each target and check if its stamp is up to date. If so,
+    // we can safely skip it. Otherwise, we must invoke the associated rule.
+    for (Target target in ordered) {
+      final List<FileSystemEntity> inputs = target.resolveInputs(environment);
+      final _InvocationEvaluation evaluation = target._canSkipInvocation(inputs, environment);
+
+      if (evaluation.canSkip) {
+        printTrace('Skipping target: ${target.name}');
+        continue;
+      }
+      printTrace('${target.name}: Starting');
+      await target.invocation(inputs, environment);
+
+      printTrace('${target.name}: Complete');
+      final List<FileSystemEntity> outputs = target.resolveOutputs(environment);
+      target.writeStamp(inputs, outputs, environment, evaluation.shas);
+    }
+  }
+
+  /// Describe the target `name` and all of its dependencies.
+  List<Map<String, Object>> describe(
+    String name,
+    Environment environment,
+  ) {
+    // Compute the required order of targets.
+    final List<Target> ordered = _computeTargetOrder(targets, name, environment);
+    final List<Map<String, Object>> result = <Map<String, Object>>[];
+    for (Target target in ordered) {
+      result.add(target.toJson(environment));
+    }
+    return result;
+  }
+
+  static List<Target> _computeTargetOrder(List<Target> allTargets, String name, Environment environment) {
```

Comment:
It will likely be useful to have a generic fold function over the dependency graph like:
```dart
class Target {
...
  T fold<T>(T initialValue, T combine(T previousValue, Target target)) {
    T depResult = target.dependencies.fold(initialValue, (T prev, Target t) => t.fold(prev, combine));
    return combine(depResult, target);
  }
...
}
```

Or a more traditional OO-style visitor thingy.

You can use something like that to:
1. Check for cycles
2. Print a target
3. Run the actions in sequence or parallel
4. etc.